src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -380,6 +393,66 @@ class NBDServer
     dout(20) << __func__ << ": terminated" << dendl;
   }
 
+  bool wait_quiesce() {
+    dout(20) << __func__ << dendl;
+
+    std::unique_lock locker{lock};
+    cond.wait(locker, [this] { return quiesce || terminated; });
+
+    if (terminated) {
+      return false;
+    }
+
+    dout(20) << __func__ << ": got quiesce request" << dendl;
+    return true;
+  }
+
+  void wait_unquiesce() {
+    dout(20) << __func__ << dendl;
+
+    std::unique_lock locker{lock};
+    cond.wait(locker, [this] { return !quiesce || terminated; });
+
+    dout(20) << __func__ << ": got unquiesce request" << dendl;
+  }
+
+  void wait_inflight_io() {
+    uint64_t features = 0;
+    image.features(&features);
+    if ((features & RBD_FEATURE_EXCLUSIVE_LOCK) != 0) {
+      bool is_owner = false;
+      image.is_exclusive_lock_owner(&is_owner);
+      if (!is_owner) {
+        return;
+      }
+    }
+
+    dout(20) << __func__ << dendl;
+
+    int r = image.flush();
```

Comment:
@dillaman There is an issue with this `flush` I forgot to mention. I think it may abuse `rbd_cache_writethrough_until_flush`. Do you think it is the problem we need to solve? Do you see a way?

Reply:
Probably should avoid the flush if you haven't seen a flush  -- or perhaps hide it behind the librbd API (i.e. after you get the `quiesce_complete` callback, kick an internal flush).

Reply:
It seems currently the rbd-nbd uses only librbd API functions and I would not want to break this until it is really necessary.

Reply:
I'm probably not understanding. I'm saying you probably should  (1) not issue the `flush` unless you are tracking if `rbd-nbd` has already seen a flush or (2) or just issue an internal flush within librbd's `quiesce_complete` call.

Reply:
I mean, how can I tell from rbd-nbd, just using librbd API, if `flush` is already issued or issue "internal" flush? It seems I need to use internal functions for this (which I really would not want) or extend the API.

And I am not sure about option (1) -- what I want here is to make sure all my pending io is complete, the `flush` call just looked like the easiest way. Not issuing `flush` I just would not know if (when) it is complete.

Reply:
You'd have to have `rbd-nbd` watch for an incoming flush and just flip a bit to know that flushes are safe. The nice thing about the internal flush is that it won't enable writeback since it's flagged as an internal-source (and therefore no different than a `block_writes` call that already happens upon snapshot create).

Reply:
> You'd have to have rbd-nbd watch for an incoming flush and just flip a bit to know that flushes are safe. 

Indeed. Though I am still not very happy with this solution, as it does not look like exactly what I want here.

> The nice thing about the internal flush is that it won't enable writeback since it's flagged as an internal-source (and therefore no different than a block_writes call that already happens upon snapshot create).

Yes, this is what I want but via API. I.e. something like `int rbd_flush2(bool dont_abuse_writethrough_until_flush)` but adding it just for this purposes looks like overkill.

I think I will do the option (1) (which is not ideal but easy to implement).

Reply:
Just for my curiosity, what's your objection to option 2?

Reply:
For the option 2 (issue internal flash) I would need to use librbd internal functions, which I reluctant to do, because currently rbd-nbd uses only API functions. Well, it uses many things from common and global that makes it impossible to ship it independently (I think), so using an internal function would not change much. But there are other points why it is nice to keep it to use "pure" librbd API: it may serve as an example for other third party applications how to use librbd and actually we know ourselves what librbd functionality is useful and what is missing. So for the option (2) we miss API to call the flush as if it is flagged as an internal-source, i.e. without enabling the writeback cache. We could add this functionality but I am not sure it is worth doing for this case. But if you think it is useful I will add `flush2` or something like this.

Reply:
Sorry, I was saying tweak the implementation of the `quiesce_complete` API call to (internally) issue an internal flush.

Reply:
Ah, interesting. Though I am not sure I like this much. I am not sure it would be obvious for users what `quiesce_complete` does and for. And I am not sure users would always want an internal flush is issued there. 

(Actually, I think even for rbd-nbd this flush is not necessary if the hook works properly, because I expect the hook (fsfreeze) would call it and wait for complete. It is here just as a safety guard, just in case the hook misbehaves.) 

And adding `sync` to `quiesce_complete` would mean `quiesce_complete` might block and I suppose we would need async version of it too (this is not need for rbd-nbd right now because it has a separate thread for it, but it might change).

Reply:
Fair enough.

Reply:
Ok. I added a commit that skips flush if it is unsafe to call it. If you still prefer a different solution please let me know.